types/src/v0/impls/l1.rs

```diff
@@ -316,14 +321,69 @@ impl L1Client {
     /// The L1 client will still be usable, but will stop updating until [`start`](Self::start) is
     /// called again.
     pub async fn shut_down_tasks(&self) {
-        if let Some(update_task) = self.update_task.0.lock().await.take() {
-            update_task.abort();
+        if let Some(mut update_task) = self.update_task.0.lock().await.take() {
+            // TODO review order
+            update_task.abort_all();
         }
     }
 
     pub fn provider(&self) -> &impl Middleware<Error: 'static> {
         &self.provider
     }
+    /// Update stake-table cache on `L1Event::NewHead`.
+    fn stake_update_loop(&self) -> impl Future<Output = ()> {
+        let opt = self.options();
+        let retry_delay = opt.l1_retry_delay;
+        let chunk_size = opt.l1_events_max_block_range as usize;
+        let span = tracing::warn_span!("L1 client update");
+        let state = self.state.clone();
+        let stake_table_contract =
+            // TODO attach address to L1Client
+            PermissionedStakeTable::new(Address::default(), self.provider.clone());
+
+        let mut events = self.receiver.activate_cloned();
+
+        async move {
+            loop {
+                let last_head = {
+                    let state = state.lock().await;
+                    state.snapshot.head
+                };
+                while let Some(event) = events.next().await {
+                    let L1Event::NewHead { head } = event else {
+                        continue;
+                    };
+
+                    let chunks = L1Client::chunky2(last_head, head, chunk_size);
+                    let mut events: Vec<StakersUpdatedFilter> = Vec::new();
+                    for (from, to) in chunks {
+                        tracing::debug!(from, to, "fetch stake table events in range");
+                        match stake_table_contract
+                            .stakers_updated_filter()
+                            .from_block(from)
+                            .to_block(to)
+                            .query()
+                            .await
+                        {
+                            Ok(e) => {
+                                for event in e {
+                                    events.push(event)
```

Comment:
Here we put the events into a local Vec called `events`. Maybe I'm missing something but it looks to me like we are not doing anything with this `events` variable.